pkgs/by-name/sp/spotify-cli/package.nix

```diff
@@ -0,0 +1,33 @@
+{ lib, python3Packages, fetchFromGitHub, buildPythonApplication }:
+
+buildPythonApplication rec {
```

Comment:
Initially I copied what the `[spotify-cli-linux`](https://github.com/NixOS/nixpkgs/blob/nixos-unstable/pkgs/applications/audio/spotify-cli-linux/default.nix) package did, I've now looked into packages that build with `python3.pkgs.buildPythonApplication` and the first one I found is also listed in `all-packages.nix`.

I'll follow your suggestion but how certain are you that it's the correct way? Why should this package, be under the `pythonPackages` namespace, and not at the top level like `spotify-cli-linux`?

Also, what about the placement in the `pkgs/by-name` hierarchy? Is it this new? Perhaps `spotify-cli-linux` should also be there? This is a bit confusion, and I want to understand, if you have any pointers!

Reply:
So, small error in my suggestion: It's preferred to use `python3Packages` instead of `python3.pkgs` for [buggy splicing reasons][splice] that I don't quite grasp. Updated suggestion:

```suggestion
{ lib, python3Packages, fetchFromGitHub }:

python3Packages.buildPythonApplication rec {
```

As for your questions: I'm fairly confident in what I'm saying, but I'll lay out my reasoning so you can double-check it! Always happy to clarify, it helps us both learn.

Definitionally, the `python3Packages` set is going to contain whatever packages we `callPackage` on in `pkgs/top-level/python-packages.nix`. Implementing my suggestion does not put the package under the `python3Packages` namespace. It simply removes the requirement to override the `buildPythonApplication` input in `pkgs/top-level/all-packages.nix`.

Since we'd no longer need that override, we can remove the `spotify-cli` entry in `pkgs/top-level/all-packages.nix`. This is possible because `pkgs/by-name` packages are called automatically by `pkgs/top-level/by-name-overlay.nix` and added to the complete package set in `pkgs/top-level/stage.nix`. I'd recommend checking out the [relevant RFC's "Semantics" section (and the RFC as a whole, tbh)][rfc] and [the implementation commit][p1-pr] if you're interested in diving deeper. 

The `pkgs/by-name` hierarchy is fairly new (see #258650). The [requirement for new packages to be located there when possible][new] is about three weeks old. If `spotify-cli-linux` were added today, it would belong in that hierarchy. We can double check that by seeing [`spotify-cli-linux` has been moved to the new hierarchy in the draft mass migration branch][mass] (albeit under a bit of a weird path, probably for testing reasons).

Hope that helps!

[splice]:https://github.com/NixOS/nixpkgs/issues/211340
[new]:https://discourse.nixos.org/t/ci-will-soon-enforce-pkgs-by-name-for-new-packages/38098
[mass]:https://github.com/nixpkgs-architecture/nixpkgs/blob/3a73796bf2edb1dc026257da827678117ee7af57/pkgs/unit/sp/spotify-cli-linux/pkg-fun.nix
[rfc]:https://github.com/NixOS/rfcs/blob/master/rfcs/0140-simple-package-paths.md#semantics
[p1-pr]:https://github.com/NixOS/nixpkgs/commit/f6467c357419d70d8f32816fe68b9bde6278f8b0

Reply:
It does, thank you!

Reply:
If you've made this change (I think so, since you marked this as resolved?), you haven't pushed it. No worries if you haven't yet, just wanted to make sure any changes don't get left behind!

Reply:
Yep! just did now